reference/config_update.md
# Configuration for auto update

Configuration for auto update

## Usage

``` r
config_update(
  series_animate = TRUE,
  update_options = FALSE,
  options_animate = TRUE,
  options_redrawPaths = TRUE,
  update_synced_charts = FALSE
)
```

## Arguments

- series_animate:

  Should the chart animate on re-rendering.

- update_options:

  Update or not global options for chart.

- options_animate:

  Should the chart animate on re-rendering.

- options_redrawPaths:

  When the chart is re-rendered, should it draw from the existing paths
  or completely redraw the chart paths from the beginning. By default,
  the chart is re-rendered from the existing paths.

- update_synced_charts:

  All the charts in a group should also update when one chart in a group
  is updated.


reference/ax_plotOptions.md
# Specific options for chart

Specific options for chart

## Usage

``` r
ax_plotOptions(
  ax,
  bar = NULL,
  heatmap = NULL,
  radialBar = NULL,
  pie = NULL,
  bubble = NULL,
  boxPlot = NULL,
  ...
)
```

## Arguments

- ax:

  An
  [`apexchart()`](https://dreamrs.github.io/apexcharter/reference/apexchart.md)
  `htmlwidget` object.

- bar:

  See
  [`bar_opts()`](https://dreamrs.github.io/apexcharter/reference/bar_opts.md).

- heatmap:

  See
  [`heatmap_opts()`](https://dreamrs.github.io/apexcharter/reference/heatmap_opts.md).

- radialBar:

  See
  [`radialBar_opts()`](https://dreamrs.github.io/apexcharter/reference/radialBar_opts.md).

- pie:

  See
  [`pie_opts()`](https://dreamrs.github.io/apexcharter/reference/pie_opts.md).

- bubble:

  See
  [`bubble_opts()`](https://dreamrs.github.io/apexcharter/reference/bubble_opts.md).

- boxPlot:

  See
  [`boxplot_opts()`](https://dreamrs.github.io/apexcharter/reference/boxplot_opts.md).

- ...:

  Additional parameters.

## Value

An
[`apexchart()`](https://dreamrs.github.io/apexcharter/reference/apexchart.md)
`htmlwidget` object.

## Examples

``` r
data("diamonds", package = "ggplot2")

# Stack bar type
apex(
  data = diamonds,
  mapping = aes(x = cut)
) %>%
  ax_plotOptions(
    bar = bar_opts(endingShape = "rounded", columnWidth = "10%")
  )

{"x":{"ax_opts":{"chart":{"type":"bar"},"series":[{"name":[],"type":"bar","data":[{"x":"Fair","y":1610},{"x":"Good","y":4906},{"x":"Very Good","y":12082},{"x":"Premium","y":13791},{"x":"Ideal","y":21551}]}],"dataLabels":{"enabled":false},"plotOptions":{"bar":{"horizontal":false,"isDumbbell":false,"endingShape":"rounded","columnWidth":"10%"}},"tooltip":{"shared":true,"intersect":false,"followCursor":true},"grid":{"yaxis":{"lines":{"show":true}},"xaxis":{"lines":{"show":false}}},"yaxis":{"labels":{"style":{"colors":"#848484"}}},"xaxis":{"labels":{"style":{"colors":"#848484"}}}},"auto_update":{"series_animate":true,"update_options":false,"options_animate":true,"options_redrawPaths":true,"update_synced_charts":false},"sparkbox":false,"xaxis":{"min":"Fair","max":"Very Good"},"type":"column"},"evals":[],"jsHooks":[]}
# Pie
apex(
  data = diamonds,
  mapping = aes(x = cut), 
  type = "pie"
) %>%
  ax_plotOptions(
    pie = pie_opts(customScale = 0.5)
  )

{"x":{"ax_opts":{"chart":{"type":"pie"},"series":[1610,4906,12082,13791,21551],"labels":["Fair","Good","Very Good","Premium","Ideal"],"yaxis":{"labels":{"style":{"colors":"#848484"}}},"xaxis":{"labels":{"style":{"colors":"#848484"}}},"plotOptions":{"pie":{"customScale":0.5}}},"auto_update":{"series_animate":true,"update_options":false,"options_animate":true,"options_redrawPaths":true,"update_synced_charts":false},"sparkbox":false,"xaxis":{"min":"Fair","max":"Very Good"},"type":"pie"},"evals":[],"jsHooks":[]}

# Radial
apexchart() %>% 
  ax_chart(type = "radialBar") %>% 
  ax_plotOptions(
    radialBar = radialBar_opts(
      hollow = list(size = "70%")
    )
  ) %>% 
  ax_series(70) %>% 
  ax_labels("Indicator")

{"x":{"ax_opts":{"chart":{"type":"radialBar"},"plotOptions":{"radialBar":{"hollow":{"size":"70%"}}},"series":[70],"labels":["Indicator"]},"auto_update":{"series_animate":true,"update_options":false,"options_animate":true,"options_redrawPaths":true,"update_synced_charts":false},"sparkbox":false},"evals":[],"jsHooks":[]}
```


NEWS.md
apexcharter 0.4.5
==================

* Updated ApexCharts.js to 5.3.6 (see https://github.com/apexcharts/apexcharts.js/releases).
* `format_num()`: added `na_label` argument to specify how to represent missing values.



apexcharter 0.4.4
==================

* Updated ApexCharts.js to 3.53.0 (see https://github.com/apexcharts/apexcharts.js/releases).
* `ax_facet_wrap`: fixed a bug generating CSS grid when using an xaxis title.



apexcharter 0.4.3
==================

* Updated ApexCharts.js to 3.49.1 (see https://github.com/apexcharts/apexcharts.js/releases).
* New chart type : slope charts.



apexcharter 0.4.2
==================

* Updated ApexCharts.js to 3.46.0 (see https://github.com/apexcharts/apexcharts.js/releases).



apexcharter 0.4.1
==================

* Updated ApexCharts.js to 3.41.0 (new charts type: dumbbell chart and funnel chart).
* `apex()` : added support for boxplot.
* New function `ax_forecast_data_points()` to mark points as forecasted values.



apexcharter 0.4.0
==================

* Updated ApexCharts.js to 3.36.3.
* New chart type : range area charts.
* Facets: correctly manage secondary y axis.



apexcharter 0.3.1
==================

* Updated ApexCharts.js to 3.33.1
* Minimal {htmlwidgets} version required >= 1.5.3



apexcharter 0.3.0
==================

* Updated ApexCharts.js to 3.29.0
* Internal: use [{packer}](https://github.com/JohnCoene/packer) to manage JavaScript assets.
* `d3.format` JavaScript functions are now available in browser under `format()` and `formatLocale()`.



apexcharter 0.2.0
==================

* Updated ApexCharts.js to 3.26.2
* New functions `ax_facet_wrap()` and `ax_facet_grid()` to create faceting charts.
* New function `apex_grid()` to combine several charts in a grid.



apexcharter 0.1.8
==================

* Updated ApexCharts.js to 3.22.2

## Bugfixes

* Fixed bad JavaScript namespace
* Fixed bug in groups with scatter chart



apexcharter 0.1.7
==================

* Updated ApexCharts.js to 3.22.0
* New chart type: treemap, see vignette for example.
* New function `ax_colors_manual()` to set color mapping manually.
* `apex()` now accept `polarArea` as type of chart.



apexcharter 0.1.6
==================

* Updated ApexCharts.js to 3.20.1
* New functions `add_line()` and `add_smooth_line()` to add simple or trend line on charts (scatter & bars).
* New Shiny input: export, to retrieve chart's base64 dataURI.



apexcharter 0.1.5
==================

* Updated ApexCharts.js to 3.18.1
* Support for candlestick charts in `apex()`.
* `apex()` has a new argument `synchronize` to easily synchronize charts together.
* `apex()` has new charts type: area-spline, area-step and step.


## New functions

* `spark_box` to create boxes with sparkline, see corresponding vignette for more details.
* `add_shade()`, `add_shade_weekend()`, `add_event()` to add annotations on time-series charts.
* `add_hline()`, `add_vline()`, `add_point()` to add annotations on charts.
* `set_tooltip_fixed()` to fix tooltip in specific position.

## Bugfixes

* Xaxis datetime now display properly with columns and bars.
* Dark mode wasn't activated properly in `ax_theme()`.




apexcharter 0.1.4
==================

* Upgraded ApexCharts.js to 3.17.1
* Fixed a bug in grouped bar charts with different levels in groups.
* New vignette to explain how to sync charts.
* New vignette to show shiny usage.
* Added functions `set_input_click()`, `set_input_zoom()` and `set_input_selection()` to add interaction in Shiny applications.



apexcharter 0.1.3
==================

* Upgraded ApexCharts.js to 3.10.1
* New function `format_num()` to format labels on y-axis or tooltip for example.
* Added localization configs file, see `?ax_chart` for examples.



apexcharter 0.1.2
==================

* Upgraded ApexCharts.js to 3.8.2
* Set parent container height to 0 by default (fix [#2](https://github.com/dreamRs/apexcharter/issues/2)).



apexcharter 0.1.1
==================

* First CRAN release: create interactive chart with the JavaScript 'ApexCharts' library. Types of graphics available : bars, columns, splines, lines, scatter, pie, donuts, heatmap, gauge.



inst/htmlwidgets/lib/Options.js

/**
 * ApexCharts Options for setting the initial configuration of ApexCharts.
 **/

var xAxisAnnotation = {
  x: 0,
  strokeDashArray: 4,
  borderColor: '#c2c2c2',
  offsetX: 0,
  offsetY: 0,
  label: {
    borderColor: '#c2c2c2',
    borderWidth: 1,
    text: "undefined",
    textAnchor: 'middle',
    orientation: 'vertical',
    position: 'top',
    offsetX: 0,
    offsetY: 0,
    style: {
      background: '#fff',
      color: '#777',
      fontSize: '12px',
      cssClass: 'apexcharts-xaxis-annotation-label',
      padding: {
        left: 5,
        right: 5,
        top: 2,
        bottom: 2
      }
    }
  }
}

var pointAnnotation = {
  x: 0,
  y: null,
  yAxisIndex: 0,
  seriesIndex: 0,
  marker: {
    size: 0,
    fillColor: '#fff',
    strokeWidth: 2,
    strokeColor: '#333',
    shape: 'circle',
    radius: 2
  },
  label: {
    borderColor: '#c2c2c2',
    borderWidth: 1,
    text: "undefined",
    textAnchor: 'middle',
    offsetX: 0,
    offsetY: -15,
    style: {
      background: '#fff',
      color: '#777',
      fontSize: '12px',
      cssClass: 'apexcharts-point-annotation-label',
      padding: {
        left: 5,
        right: 5,
        top: 0,
        bottom: 2
      }
    }
  }
}

var yAxisAnnotation = {
  y: 0,
  strokeDashArray: 4,
  borderColor: '#c2c2c2',
  offsetX: 0,
  offsetY: 0,
  yAxisIndex: 0,
  label: {
    borderColor: '#c2c2c2',
    borderWidth: 1,
    text: "undefined",
    textAnchor: 'end',
    position: 'right',
    offsetX: 0,
    offsetY: -3,
    style: {
      background: '#fff',
      color: '#777',
      fontSize: '12px',
      cssClass: 'apexcharts-yaxis-annotation-label',
      padding: {
        left: 5,
        right: 5,
        top: 0,
        bottom: 2
      }
    }
  }
}

var yAxis = {
  opposite: false,
  tickAmount: 6,
  max: "undefined",
  min: "undefined",
  floating: false,
  labels: {
    show: true,
    maxWidth: 160,
    offsetX: 0,
    offsetY: 0,
    style: {
      colors: [],
      fontSize: '12px',
      cssClass: 'apexcharts-yaxis-label'
    },
    formatter: function (val) {
      if (typeof val === 'number') {
        return val ? val.toFixed(0) : val
      }
      return val
    }
  },
  axisBorder: {
    show: false,
    color: '#78909C',
    offsetX: 0,
    offsetY: 0
  },
  axisTicks: {
    show: false,
    color: '#78909C',
    width: 6,
    offsetX: 0,
    offsetY: 0
  },
  title: {
    text: "undefined",
    rotate: -90,
    offsetY: 0,
    offsetX: 0,
    style: {
      color: "undefined",
      fontSize: '12px',
      cssClass: 'apexcharts-yaxis-title'
    }
  },
  tooltip: {
    enabled: false,
    offsetX: 0
  },
  crosshairs: {
    show: true,
    position: 'front',
    stroke: {
      color: '#b6b6b6',
      width: 1,
      dashArray: 0
    }
  }
}

var Options = {
  //init () {
  //  return {
      annotations: {
        position: 'front',
        yaxis: [yAxisAnnotation],
        xaxis: [xAxisAnnotation],
        points: [pointAnnotation]
      },
      chart: {
        animations: {
          enabled: true,
          easing: 'easeinout', // linear, easeout, easein, easeinout
          speed: 800,
          animateGradually: {
            delay: 150,
            enabled: true
          },
          dynamicAnimation: {
            enabled: true,
            speed: 350
          }
        },
        background: 'transparent',
        foreColor: '#373d3f',
        dropShadow: {
          enabled: false,
          enabledSeries: "undefined",
          top: 2,
          left: 2,
          blur: 4,
          opacity: 0.35
        },
        events: {
          beforeMount: "undefined",
          mounted: "undefined",
          updated: "undefined",
          clicked: "undefined",
          selection: "undefined",
          dataPointSelection: "undefined",
          zoomed: "undefined",
          scrolled: "undefined"
        },
        height: 'auto',
        offsetX: 0,
        offsetY: 0,
        scroller: {
          enabled: false,
          height: 30,
          track: {
            height: 2,
            background: '#e0e0e0'
          },
          thumb: {
            height: 2,
            background: '#008FFB'
          },
          scrollButtons: {
            enabled: true,
            size: 6,
            borderWidth: 2,
            borderColor: '#c3c3c3',
            fillColor: '#fff'
          },
          padding: {
            left: 10,
            right: 10
          },
          offsetX: 0,
          offsetY: 0
        },
        selection: {
          enabled: true,
          type: 'x',
          selectedPoints: "undefined",
          fill: {
            color: '#24292e',
            opacity: 0.1
          },
          stroke: {
            width: 1,
            color: '#24292e',
            opacity: 0.4,
            dashArray: 3
          },
          xaxis: {
            min: "undefined",
            max: "undefined"
          },
          yaxis: {
            min: "undefined",
            max: "undefined"
          }
        },
        sparkline: {
          enabled: false
        },
        stacked: false,
        stackType: 'normal',
        toolbar: {
          show: true,
          tools: {
            download: true,
            selection: true,
            zoom: true,
            zoomin: true,
            zoomout: true,
            pan: true,
            reset: true
          }
        },
        type: 'line',
        width: '100%',
        zoom: {
          enabled: true,
          type: 'x',
          zoomedArea: {
            fill: {
              color: '#90CAF9',
              opacity: 0.4
            },
            stroke: {
              color: '#0D47A1',
              opacity: 0.4,
              width: 1
            }
          }
        }
      },
      plotOptions: {
        bar: {
          horizontal: false,
          endingShape: 'flat',
          columnWidth: '70%', // should be in percent 0 - 100
          barHeight: '70%', // should be in percent 0 - 100
          distributed: false,
          colors: {
            ranges: [],
            backgroundBarColors: [],
            backgroundBarOpacity: 1
          },
          dataLabels: {
            position: 'top' // top, center, bottom
          }
          // stackedLabels: true
        },
        heatmap: {
          radius: 2,
          enableShades: true,
          shadeIntensity: 0.5,
          colorScale: {
            ranges: []
          }
        },
        radialBar: {
          size: "undefined",
          inverseOrder: false,
          startAngle: 0,
          endAngle: 360,
          offsetX: 0,
          offsetY: 0,
          hollow: {
            margin: 5,
            size: '50%',
            background: 'transparent',
            image: "undefined",
            imageWidth: 150,
            imageHeight: 150,
            imageOffsetX: 0,
            imageOffsetY: 0,
            imageClipped: true,
            position: 'front',
            dropShadow: {
              enabled: false,
              top: 0,
              left: 0,
              blur: 3,
              opacity: 0.5
            }
          },
          track: {
            show: true,
            startAngle: "undefined",
            endAngle: "undefined",
            background: '#f2f2f2',
            strokeWidth: '97%',
            opacity: 1,
            margin: 5, // margin is in pixels
            dropShadow: {
              enabled: false,
              top: 0,
              left: 0,
              blur: 3,
              opacity: 0.5
            }
          },
          dataLabels: {
            showOn: 'always', // hover/always
            name: {
              show: true,
              fontSize: '22px',
              color: "undefined",
              offsetY: -10
            },
            value: {
              show: true,
              fontSize: '16px',
              color: "undefined",
              offsetY: 16,
              formatter: function (val) {
                return val + '%'
              }
            }
          }
        },
        pie: {
          size: "undefined",
          donut: {
            size: '65%',
            background: 'transparent'
            // TODO: draw labels in donut area
            // labels: {
            //   showOn: 'hover',
            //   name: {
            //     show: false,
            //     fontSize: '22px',
            //     color: "undefined",
            //     offsetY: -10
            //   },
            //   value: {
            //     show: true,
            //     offsetY: 16,
            //     fontSize: '16px',
            //     color: "undefined",
            //     formatter: function (val) {
            //       return val + '%'
            //     }
            //   }
            // }
          },
          customScale: 0,
          offsetX: 0,
          offsetY: 0,
          dataLabels: {
            offset: 0 // offset by which labels will move outside
          }
        }
      },
      colors: "undefined",
      dataLabels: {
        enabled: true,
        formatter: function (val) {
          return val
        },
        textAnchor: 'middle',
        offsetX: 0,
        offsetY: 0,
        style: {
          fontSize: '14px',
          colors: "undefined"
        },
        dropShadow: {
          enabled: false,
          top: 1,
          left: 1,
          blur: 1,
          opacity: 0.45
        }
      },
      fill: {
        type: 'solid',
        colors: "undefined", // array of colors,
        opacity: 0.9,
        gradient: {
          shade: 'dark',
          type: 'horizontal',
          shadeIntensity: 0.5,
          gradientToColors: "undefined",
          inverseColors: true,
          opacityFrom: 1,
          opacityTo: 1,
          stops: [0, 50, 100]
        },
        image: {
          src: [],
          width: "undefined", // optional
          height: "undefined" // optional
        },
        pattern: {
          style: 'sqaures', // string or array of strings
          width: 6,
          height: 6,
          strokeWidth: 2
        }
      },
      grid: {
        show: true,
        borderColor: '#e0e0e0',
        strokeDashArray: 0,
        position: 'back',
        xaxis: {
          lines: {
            show: false,
            animate: false
          }
        },
        yaxis: {
          lines: {
            show: true,
            animate: true
          }
        },
        row: {
          colors: "undefined", // takes as array which will be repeated on rows
          opacity: 0.5
        },
        column: {
          colors: "undefined", // takes an array which will be repeated on columns
          opacity: 0.5
        },
        padding: {
          top: 0,
          right: 10,
          bottom: 0,
          left: 10
        }
      },
      labels: [],
      legend: {
        show: true,
        floating: false,
        position: 'bottom', // whether to position legends in 1 of 4
        // direction - top, bottom, left, right
        horizontalAlign: 'center', // when position top/bottom, you can
        // specify whether to align legends
        // left, right or center
        verticalAlign: 'middle',
        fontSize: '14px',
        textAnchor: 'start',
        offsetY: 0,
        offsetX: 0,
        formatter: "undefined",
        labels: {
          color: "undefined",
          useSeriesColors: false
        },
        markers: {
          size: 6,
          strokeWidth: 0,
          strokeColor: '#fff',
          offsetX: 0,
          offsetY: 0,
          shape: 'circle',
          radius: 2
        },
        itemMargin: {
          horizontal: 20,
          vertical: 5
        },
        containerMargin: {
          left: 5,
          top: 4,
          right: 0,
          bottom: 0
        },
        onItemClick: {
          toggleDataSeries: true
        },
        onItemHover: {
          highlightDataSeries: true
        }
      },
      markers: {
        discrete: [],
        size: 0,
        colors: "undefined",
        strokeColor: '#fff',
        strokeWidth: 2,
        strokeOpacity: 0.9,
        fillOpacity: 1,
        shape: 'circle',
        radius: 2,
        offsetX: 0,
        offsetY: 0,
        hover: {
          size: 6
        }
      },
      noData: {
        text: "undefined",
        align: 'center',
        verticalAlign: 'middle',
        offsetX: 0,
        offsetY: 0,
        style: {
          color: '#888',
          fontSize: '16px'
        }
      },
      responsive: [], // breakpoints should follow ascending order 400, then 700, then 1000
      series: "undefined",
      states: {
        normal: {
          filter: {
            type: 'none',
            value: 0
          }
        },
        hover: {
          filter: {
            type: 'lighten',
            value: 0.15
          }
        },
        active: {
          allowMultipleDataPointsSelection: false,
          filter: {
            type: 'darken',
            value: 0.35
          }
        }
      },
      title: {
        text: "undefined",
        align: 'left',
        margin: 10,
        offsetX: 0,
        offsetY: 0,
        floating: false,
        style: {
          fontSize: '16px',
          color: '#263238'
        }
      },
      subtitle: {
        text: "undefined",
        align: 'left',
        margin: 10,
        offsetX: 0,
        offsetY: 0,
        floating: false,
        style: {
          fontSize: '14px',
          color: '#9699a2'
        }
      },
      stroke: {
        show: true,
        curve: 'smooth', // "smooth" or "straight"
        lineCap: 'butt', // round, butt , square
        width: 2,
        colors: "undefined", // array of colors
        dashArray: 0 // single value or array of values
      },
      tooltip: {
        enabled: true,
        shared: true,
        followCursor: false, // when disabled, the tooltip will show on top of the series instead of mouse position
        intersect: false, // when enabled, tooltip will only show when user directly hovers over point
        inverseOrder: false,
        custom: "undefined",
        fillSeriesColor: false,
        onDatasetHover: {
          highlightDataSeries: false
        },
        theme: 'light',
        x: { // x value
          show: true,
          format: 'dd MMM', // dd/MM, dd MMM yy, dd MMM yyyy
          formatter: "undefined" // a custom user supplied formatter function
        },
        y: {
          formatter: "undefined",
          title: {
            formatter: function (seriesName) {
              return seriesName
            }
          }
        },
        z: {
          formatter: "undefined",
          title: 'Size: '
        },
        marker: {
          show: true
        },
        items: {
          display: 'flex'
        },
        fixed: {
          enabled: false,
          position: 'topRight', // topRight, topLeft, bottomRight, bottomLeft
          offsetX: -100,
          offsetY: 0
        }
      },
      xaxis: {
        type: 'category',
        categories: [],
        labels: {
          show: true,
          rotate: -45,
          rotateAlways: false,
          trim: true,
          maxHeight: 120,
          style: {
            colors: [],
            fontSize: '12px',
            cssClass: 'apexcharts-xaxis-label'
          },
          offsetX: 0,
          offsetY: 0,
          format: "undefined",
          formatter: "undefined", // custom formatter function which will override format
          datetimeFormatter: {
            year: 'yyyy',
            month: 'MMM \'yy',
            day: 'dd MMM',
            hour: 'HH:mm'
          }
        },
        axisBorder: {
          show: true,
          color: '#78909C',
          offsetX: 0,
          offsetY: 0
        },
        axisTicks: {
          show: true,
          color: '#78909C',
          height: 6,
          offsetX: 0,
          offsetY: 0
        },
        tickAmount: "undefined",
        min: "undefined",
        max: "undefined",
        range: "undefined",
        floating: false,
        position: 'bottom',
        title: {
          text: "undefined",
          offsetX: 0,
          offsetY: 0,
          style: {
            color: "undefined",
            fontSize: '12px',
            cssClass: 'apexcharts-xaxis-title'
          }
        },
        crosshairs: {
          show: true,
          width: 1, // tickWidth/barWidth or an integer
          position: 'back',
          opacity: 0.9,
          stroke: {
            color: '#b6b6b6',
            width: 0,
            dashArray: 0
          },
          fill: {
            type: 'solid', // solid, gradient
            color: '#B1B9C4',
            gradient: {
              colorFrom: '#D8E3F0',
              colorTo: '#BED1E6',
              stops: [0, 100],
              opacityFrom: 0.4,
              opacityTo: 0.5
            }
          },
          dropShadow: {
            enabled: false,
            left: 0,
            top: 0,
            blur: 1,
            opacity: 0.4
          }
        },
        tooltip: {
          enabled: true,
          offsetY: 0
        }
      },
      yaxis: yAxis,
      theme: {
        palette: 'palette1', // If defined, it will overwrite config.colors variable
        monochrome: { // monochrome allows you to select just 1 color and fill out the rest with light/dark shade (intensity can be selected)
          enabled: false,
          color: '#008FFB',
          shadeTo: 'light',
          shadeIntensity: 0.65
        }
      }
  //  }
  //}
}
